src/resolver/environment.cc

```diff
@@ -148,37 +149,51 @@ Environment::groupIsLocked( const InstallDescriptors & group,
 {
   auto packages = oldLockfile.getLockfileRaw().packages;
   if ( ! packages.contains( system ) ) { return false; }
+
   SystemPackages oldSystemPackages = packages.at( system );
 
   InstallDescriptors oldDescriptors = oldLockfile.getDescriptors();
 
+  /* Check for upgrades. */
   for ( auto & [iid, descriptor] : group )
     {
-      /* Check for upgrades. */
-      if ( const bool * upgradeEverything
-           = std::get_if<bool>( &this->upgrades ) )
-        {
-          /* If we're upgrading everything, the group needs to be locked again.
-           */
-          if ( *upgradeEverything ) { return false; }
-        }
-      else
-        {
-          /* If the current iid is being upgraded, the group needs to be
-           * locked again. */
-          auto upgrades = std::get<std::vector<InstallID>>( this->upgrades );
-          if ( std::find( upgrades.begin(), upgrades.end(), iid )
-               != upgrades.end() )
-            {
-              return false;
-            }
-        }
+      {
+        bool stale = false;
+        std::visit(
+          overloaded { /* If we are upgrading everything, we treat all groups
+                        * as "unlocked". */
+                       [&]( bool upgradeEverything )
+                       {
+                         if ( upgradeEverything ) { stale = true; }
+                       },
+
+                       /* If the current iid is being upgraded, the group needs
+                        * to be locked again. */
+                       [&]( std::vector<InstallID> upgrades )
+                       {
+                         if ( std::find( upgrades.begin(), upgrades.end(), iid )
+                              != upgrades.end() )
+                           {
+                             stale = true;
+                           }
+                       } },
+          this->upgrades );
+        if ( stale ) { return false; }
+      }
+
       /* If the descriptor has changed compared to the one in the lockfile
        * manifest, it needs to be locked again. */
```

Comment:
nonblocking: might move this comment or reword it as it's describing the opposite of what the conditional is checking